cmd/upload_release.go

```diff
@@ -34,52 +39,71 @@ func NewUploadReleaseCmd(
 		director:              director,
 		releaseArchiveFactory: releaseArchiveFactory,
 
-		ui: ui,
+		cmdRunner: cmdRunner,
+		fs:        fs,
+		ui:        ui,
 	}
 }
 
 func (c UploadReleaseCmd) Run(opts UploadReleaseOpts) error {
-	if opts.Release != nil {
+	switch {
+	case opts.Release != nil:
 		return c.uploadRelease(opts.Release, opts)
+	case opts.Args.URL.IsRemote():
+		return c.uploadIfNecessary(opts, c.uploadRemote)
+	case opts.Args.URL.IsGit():
+		return c.uploadIfNecessary(opts, c.uploadGit)
+	default:
+		return c.uploadFile(opts)
 	}
+}
+
+func (c UploadReleaseCmd) uploadRemote(opts UploadReleaseOpts) error {
+	return c.director.UploadReleaseURL(string(opts.Args.URL), opts.SHA1, opts.Rebase, opts.Fix)
+}
 
-	if opts.Args.URL.IsRemote() {
-		return c.uploadRemote(string(opts.Args.URL), opts)
+func (c UploadReleaseCmd) uploadGit(opts UploadReleaseOpts) error {
+	repoPath, err := c.fs.TempDir("bosh-upload-release-git-clone")
+	if err != nil {
+		return bosherr.WrapErrorf(err, "Creating tmp dir for git cloning")
 	}
 
-	return c.uploadFile(opts)
-}
+	defer c.fs.RemoveAll(repoPath)
 
-func (c UploadReleaseCmd) uploadRemote(url string, opts UploadReleaseOpts) error {
-	version := semver.Version(opts.Version)
+	_, _, _, err = c.cmdRunner.RunCommand("git", "clone", opts.Args.URL.GitRepo(), "--depth", "1", repoPath)
```

Comment:
Might want to set `GIT_TERMINAL_PROMPT=0` to avoid infinite hangs if the user typos or tries to do an otherwise interactive git clone (like on a typo'd repo or one requiring authn).